compiler/src/org.graalvm.compiler.loop.phases/src/org/graalvm/compiler/loop/phases/LoopPredicationPhase.java

```diff
@@ -0,0 +1,282 @@
+                            mainPathNodes.add(input);
+                        }
+                    }
+                }
+            }
+            if (cur == begin) {
+                break;
+            }
+            cur = cur.getDominator();
+        }
+
+        assert workList.isEmpty();
+        return mainPathNodes;
+    }
+
+    // Create a range check predicate for
+    //
+    // for (i = init; i < limit; i += stride) {
+    // a[scale*i+offset]
+    // }
+    private static void processGuard(LoopEx loop, GuardNode guard) {
+        final LogicNode condition = guard.getCondition();
+        if (!(condition instanceof IntegerBelowNode) || guard.isNegated()) {
+            return;
+        }
+
+        final IntegerBelowNode rangeCheck = (IntegerBelowNode) condition;
+        final ValueNode range = rangeCheck.getY();
+        if (!loop.isOutsideLoop(range) || ((IntegerStamp) range.stamp(NodeView.DEFAULT)).lowerBound() < 0) {
+            return;
+        }
+
+        final EconomicMap<Node, InductionVariable> inductionVariables = loop.getInductionVariables();
+
+        final ValueNode x = rangeCheck.getX();
+        if (!inductionVariables.containsKey(x)) {
+            return;
+        }
+
+        final StructuredGraph graph = guard.graph();
+
+        final InductionVariable iv = inductionVariables.get(x);
+
+        Long scale = null;
+
+        final InductionVariable counter = loop.counted().getCounter();
+        if (iv.isConstantScale(counter)) {
+            scale = iv.constantScale(counter);
+        }
+
+        ValueNode offset = null;
+
+        if (iv.offsetIsZero(counter)) {
+            offset = graph.unique(ConstantNode.forInt(0));
+        } else {
+            offset = iv.offsetNode(counter);
+        }
+
+        if (offset == null || scale == null || !loop.isOutsideLoop(offset)) {
+            return;
+        }
+
+        long scaleCon = scale;
+
+        final GuardNode upperGuard = getGuardNode(loop, guard, range, graph, scaleCon, offset, true);
+        final GuardNode lowerGuard = getGuardNode(loop, guard, range, graph, scaleCon, offset, false);
+
+        GuardingNode combinedGuard = null;
+        for (Node usage : guard.usages().snapshot()) {
+            if (usage instanceof MultiGuardNode) {
+                final MultiGuardNode multiGuardNode = (MultiGuardNode) usage;
+                multiGuardNode.replaceFirstInput(guard, lowerGuard);
+                multiGuardNode.addGuard(upperGuard);
+            } else {
+                if (combinedGuard == null) {
+                    combinedGuard = MultiGuardNode.combine(lowerGuard, upperGuard);
+                }
+                ((GuardedNode) usage).setGuard(combinedGuard);
+            }
+        }
+        assert guard.hasNoUsages();
+        guard.safeDelete();
+    }
+
+    private static GuardNode getGuardNode(LoopEx loop, GuardNode guard, ValueNode range, StructuredGraph graph, long scaleCon, ValueNode offset, boolean upper) {
```

Comment:
This could also use `initNode` and `extremumNode` from `InductionVariable` to from the upper and lower guards

Reply:
I checked the codes, and it seems initNode/extremumNode is not directly equals to "min/max" for an induction when the directly is `DOWN`.  So maybe it needs to handle the difference between "UP" and "DOWN", doesn't it?

Reply:
That's right which one is min vs. max depends on direction

Reply:
For range check:
scale * iv + offset |<| range
and lower <= iv <= upper
we need 2 guards:
scale * lower + offset |<| range
scale * upper + offset |<| range

We test both ends of the range of values taken by the iv with the same formula. There's no need to handle UP and DOWN differently as long as we indeed add a guard for each ends of the iteration range.

Reply:
Yes, indeed! Thanks!